File: data/derived_cot/rq1_traces/translation/r1/java_to_cpp/code_output/Hotel.cpp
#include <iostream>
#include <map>
#include <string>
#include <iomanip>

std::ostream& operator<<(std::ostream& os, const std::map<std::string, int>& m) {
    os << "{";
    bool first = true;
    for (const auto& p : m) {
        if (!first) {
            os << ", ";
        }
        first = false;
        os << p.first << "=" << p.second;
    }
    os << "}";
    return os;
}

std::ostream& operator<<(std::ostream& os, const std::map<std::string, std::map<std::string, int>>& m) {
    os << "{";
    bool first = true;
    for (const auto& p : m) {
        if (!first) {
            os << ", ";
        }
        first = false;
        os << p.first << "=" << p.second;
    }
    os << "}";
    return os;
}

class Hotel {
private:
    std::string name;
    std::map<std::string, int> availableRooms;
    std::map<std::string, std::map<std::string, int>> bookedRooms;

public:
    Hotel(const std::string& name, const std::map<std::string, int>& rooms)
        : name(name), availableRooms(rooms) {}

    std::string bookRoom(const std::string& roomType, int roomNumber, const std::string& guestName) {
        if (availableRooms.find(roomType) == availableRooms.end()) {
            return "False";
        }

        int available = availableRooms[roomType];
        if (roomNumber <= available) {
            bookedRooms[roomType][guestName] = roomNumber;
            availableRooms[roomType] = available - roomNumber;
            return "Success!";
        } else {
            return "False";
        }
    }

    bool checkIn(const std::string& roomType, int roomNumber, const std::string& guestName) {
        auto roomIt = bookedRooms.find(roomType);
        if (roomIt == bookedRooms.end()) {
            return false;
        }

        auto& innerMap = roomIt->second;
        auto guestIt = innerMap.find(guestName);
        if (guestIt == innerMap.end()) {
            return false;
        }

        int booked = guestIt->second;
        if (roomNumber > booked) {
            return false;
        } else if (roomNumber == booked) {
            innerMap.erase(guestIt);
        } else {
            guestIt->second = booked - roomNumber;
        }
        return true;
    }

    void checkOut(const std::string& roomType, int roomNumber) {
        availableRooms[roomType] += roomNumber;
    }

    int getAvailableRooms(const std::string& roomType) const {
        auto it = availableRooms.find(roomType);
        if (it != availableRooms.end()) {
            return it->second;
        }
        return 0;
    }

    friend int main();
};

int main() {
    std::map<std::string, int> rooms;
    rooms["single"] = 3;
    rooms["double"] = 2;
    Hotel hotel("Test Hotel", rooms);

    std::cout << std::boolalpha;

    std::cout << hotel.bookRoom("single", 2, "guest 1") << std::endl;
    std::cout << hotel.bookRoom("triple", 2, "guest 1") << std::endl;
    std::cout << hotel.bookRoom("single", 2, "guest 2") << std::endl;
    std::cout << hotel.bookRoom("single", 1, "guest 2") << std::endl;
    std::cout << hotel.bookRoom("single", 3, "guest 1") << std::endl;
    std::cout << hotel.bookRoom("single", 100, "guest 1") << std::endl;

    hotel.checkIn("single", 1, "guest 1");
    std::cout << hotel.bookedRooms << std::endl;
    std::cout << hotel.checkIn("single", 3, "guest 1") << std::endl;
    std::cout << hotel.checkIn("double", 1, "guest 1") << std::endl;
    hotel.checkIn("double", 1, "guest 2");
    std::cout << hotel.bookedRooms << std::endl;

    hotel.checkOut("single", 1);
    std::cout << hotel.availableRooms << std::endl;
    hotel.checkOut("triple", 2);
    std::cout << hotel.availableRooms << std::endl;

    std::cout << hotel.getAvailableRooms("single") << std::endl;

    return 0;
}

File: data/derived_cot/rq1_traces/translation/glm5.1/java_to_py/code_output/Hotel.py
class Hotel:
    def __init__(self, name, rooms):
        self.name = name
        self.available_rooms = dict(rooms)
        self.booked_rooms = {}

    def book_room(self, room_type, room_number, name):
        if room_type not in self.available_rooms:
            return "False"

        available = self.available_rooms[room_type]
        if room_number <= available:
            self.booked_rooms.setdefault(room_type, {})
            self.booked_rooms[room_type][name] = room_number
            self.available_rooms[room_type] = available - room_number
            return "Success!"
        else:
            return "False"

    def check_in(self, room_type, room_number, name):
        if room_type not in self.booked_rooms or name not in self.booked_rooms[room_type]:
            return False

        booked = self.booked_rooms[room_type][name]
        if room_number > booked:
            return False
        elif room_number == booked:
            del self.booked_rooms[room_type][name]
        else:
            self.booked_rooms[room_type][name] = booked - room_number
        return True

    def check_out(self, room_type, room_number):
        self.available_rooms[room_type] = self.available_rooms.get(room_type, 0) + room_number

    def get_available_rooms(self, room_type):
        return self.available_rooms.get(room_type, 0)


if __name__ == "__main__":
    rooms = {"single": 3, "double": 2}
    hotel = Hotel("Test Hotel", rooms)

    print(hotel.book_room("single", 2, "guest 1"))
    print(hotel.book_room("triple", 2, "guest 1"))
    print(hotel.book_room("single", 2, "guest 2"))
    print(hotel.book_room("single", 1, "guest 2"))
    print(hotel.book_room("single", 3, "guest 1"))
    print(hotel.book_room("single", 100, "guest 1"))

    hotel.check_in("single", 1, "guest 1")
    print(hotel.booked_rooms)
    print(hotel.check_in("single", 3, "guest 1"))
    print(hotel.check_in("double", 1, "guest 1"))
    hotel.check_in("double", 1, "guest 2")
    print(hotel.booked_rooms)

    hotel.check_out("single", 1)
    print(hotel.available_rooms)
    hotel.check_out("triple", 2)
    print(hotel.available_rooms)

    print(hotel.get_available_rooms("single"))